routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,381 @@
+      WriteToSink(sink, m_size);
+      WriteToSink(sink, m_numRoads);
+      WriteToSink(sink, m_beginJointId);
+      WriteToSink(sink, m_endJointId);
+      WriteToSink(sink, m_mask);
+    }
+
+    template <class Source>
+    void Deserialize(Source & src)
+    {
+      m_size = ReadPrimitiveFromSource<decltype(m_size)>(src);
+      m_numRoads = ReadPrimitiveFromSource<decltype(m_numRoads)>(src);
+      m_beginJointId = ReadPrimitiveFromSource<decltype(m_beginJointId)>(src);
+      m_endJointId = ReadPrimitiveFromSource<decltype(m_endJointId)>(src);
+      m_mask = ReadPrimitiveFromSource<decltype(m_mask)>(src);
+    }
+
+    uint64_t GetSize() const { return m_size; }
+    uint32_t GetNumRoads() const { return m_numRoads; }
+    Joint::Id GetBeginJointId() const { return m_beginJointId; }
+    Joint::Id GetEndJointId() const { return m_endJointId; }
+    VehicleMask GetMask() const { return m_mask; }
+
+  private:
+    uint64_t m_size = 0;
+    uint32_t m_numRoads = 0;
+    Joint::Id m_beginJointId = Joint::kInvalidId;
+    Joint::Id m_endJointId = Joint::kInvalidId;
+    VehicleMask m_mask = 0;
+  };
+
+  class Header final
+  {
+  public:
+    Header() = default;
+
+    Header(IndexGraph const & graph, unordered_map<uint32_t, VehicleMask> const & masks)
+      : m_numRoads(graph.GetNumRoads()), m_numJoints(graph.GetNumJoints())
+    {
+    }
+
+    template <class Sink>
+    void Serialize(Sink & sink) const
+    {
+      WriteToSink(sink, m_version);
+      WriteToSink(sink, m_numRoads);
+      WriteToSink(sink, m_numJoints);
+
+      WriteToSink(sink, static_cast<uint32_t>(m_sections.size()));
+      for (Section const & section : m_sections)
+        section.Serialize(sink);
+    }
+
+    template <class Source>
+    void Deserialize(Source & src)
+    {
+      m_version = ReadPrimitiveFromSource<decltype(m_version)>(src);
+      if (m_version != kVersion)
+      {
+        MYTHROW(RoutingException,
+                ("Unknown index graph version ", m_version, ", current version ", kVersion));
+      }
+
+      m_numRoads = ReadPrimitiveFromSource<decltype(m_numRoads)>(src);
+      m_numJoints = ReadPrimitiveFromSource<decltype(m_numJoints)>(src);
+
+      auto const sectionsSize = ReadPrimitiveFromSource<uint32_t>(src);
+      m_sections.resize(sectionsSize);
+      for (Section & section : m_sections)
+        section.Deserialize(src);
+    }
+
+    uint32_t GetNumRoads() const { return m_numRoads; }
+    Joint::Id GetNumJoints() const { return m_numJoints; }
+    uint32_t GetNumSections() const { return m_sections.size(); }
+
+    Section const & GetSection(uint32_t index) const
+    {
+      CHECK_LESS(index, m_sections.size(), ());
+      return m_sections[index];
+    }
+
+    void AddSection(Section const & section) { m_sections.push_back(section); }
+
+  private:
+    uint8_t m_version = kVersion;
+    uint32_t m_numRoads = 0;
+    Joint::Id m_numJoints = 0;
+    vector<Section> m_sections;
+  };
+
+  class JointIdEncoder final
```

Comment:
Ah, OK, seems that you really need JointIdEncoder. But the constructor parameter named begin in JointIdDecoder is disguising, please, rename it, because decoder may return values that are less than begin, right?